File: javascript/allergies/allergies.js
(function() {

  'use strict';

  function Allergies(score) {
    this.score = score;
    this.scores = [1, 2, 4, 8, 16, 32, 64, 128];
    this.allergies = ["eggs", "peanuts", "shellfish", 
      "strawberries", "tomatoes", "chocolate", "pollen", "cats"];
    this.scoresToAllergiesMapping = this._zipScoresAllergies();
  }

  Allergies.prototype.list = function() {
    var validScores = this.score % 256;
    var allergy;
    var currentScore = validScores;
    var availableScores = validScores;
    var list = [];
    while(currentScore > 0) {
      if((allergy = this.scoresToAllergiesMapping[currentScore]) && (availableScores - currentScore >= 0)) {
        list.push(allergy);
        availableScores -= currentScore;
      } 
      currentScore -= 1
    }
    return list.reverse();
  }

  Allergies.prototype.allergicTo = function(allergy) {
    return this.list().indexOf(allergy) !== -1;
  }

  Allergies.prototype._zipScoresAllergies = function() {
    var scoresToAllergiesMapping = {}, i = 0, currentScore, currentAllergy
    for(i; i < this.scores.length; i++) {
      currentScore = this.scores[i];
      currentAllergy = this.allergies[i]
      scoresToAllergiesMapping[currentScore] = currentAllergy;
    };
    return scoresToAllergiesMapping;
  }

  module.exports = Allergies;

})();

File: ruby/allergies/allergies.rb
class Allergies
  
  attr_reader :score, :all_allergies, :scores, :scores_allergies_mapping
  private :score, :all_allergies, :scores, :scores_allergies_mapping

  def initialize(score)
    @score = score
    @all_allergies = %w[eggs peanuts shellfish strawberries tomatoes chocolate pollen cats]
    @scores = %w[1 2 4 8 16 32 64 128].map(&:to_i)
    @scores_allergies_mapping = scores.zip(all_allergies).to_h
  end

  def list
    current_scores = score
    available_scores = score % limit

    allergies = []

    while (available_scores > 0)

      if scores.include?(current_scores) and (available_scores - current_scores >= 0)
        available_scores -= current_scores
        allergies << scores_allergies_mapping[current_scores]
      end

      current_scores -= 1

    end
    allergies.reverse
  end

  def allergic_to?(allergy)
    list.include? allergy
  end

  private

    def limit
      scores.reduce(0, :+) + 1
    end

end